File: petclinic-java-gradle-react-tailwind-ts/src/main/java/dev/pollito/petclinic_java_gradle_react_tailwind_ts/specialty/SpecialtyServiceImpl.java
package dev.pollito.petclinic_java_gradle_react_tailwind_ts.specialty;

import dev.pollito.petclinic_java_gradle_react_tailwind_ts.events.BeforeDeleteSpecialty;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.util.CustomCollectors;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.util.NotFoundException;
import java.util.List;
import java.util.Map;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(rollbackFor = Exception.class)
public class SpecialtyServiceImpl implements SpecialtyService {

    private final SpecialtyRepository specialtyRepository;
    private final ApplicationEventPublisher publisher;
    private final SpecialtyMapper specialtyMapper;

    public SpecialtyServiceImpl(final SpecialtyRepository specialtyRepository,
            final ApplicationEventPublisher publisher, final SpecialtyMapper specialtyMapper) {
        this.specialtyRepository = specialtyRepository;
        this.publisher = publisher;
        this.specialtyMapper = specialtyMapper;
    }

    @Override
    public List<SpecialtyDTO> findAll() {
        final List<Specialty> specialties = specialtyRepository.findAll(Sort.by("id"));
        return specialties.stream()
                .map(specialty -> specialtyMapper.updateSpecialtyDTO(specialty, new SpecialtyDTO()))
                .toList();
    }

    @Override
    public SpecialtyDTO get(final Integer id) {
        return specialtyRepository.findById(id)
                .map(specialty -> specialtyMapper.updateSpecialtyDTO(specialty, new SpecialtyDTO()))
                .orElseThrow(NotFoundException::new);
    }

    @Override
    public Integer create(final SpecialtyDTO specialtyDTO) {
        final Specialty specialty = new Specialty();
        specialtyMapper.updateSpecialty(specialtyDTO, specialty);
        return specialtyRepository.save(specialty).getId();
    }

    @Override
    public void update(final Integer id, final SpecialtyDTO specialtyDTO) {
        final Specialty specialty = specialtyRepository.findById(id)
                .orElseThrow(NotFoundException::new);
        specialtyMapper.updateSpecialty(specialtyDTO, specialty);
        specialtyRepository.save(specialty);
    }

    @Override
    public void delete(final Integer id) {
        final Specialty specialty = specialtyRepository.findById(id)
                .orElseThrow(NotFoundException::new);
        publisher.publishEvent(new BeforeDeleteSpecialty(id));
        specialtyRepository.delete(specialty);
    }

    @Override
    public Map<Integer, Integer> getSpecialtyValues() {
        return specialtyRepository.findAll(Sort.by("id"))
                .stream()
                .collect(CustomCollectors.toSortedMap(Specialty::getId, Specialty::getId));
    }

}

File: petclinic-kt-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/main/kotlin/dev/pollito/petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts/specialty/SpecialtyServiceImpl.kt
package dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.specialty

import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.events.BeforeDeleteSpecialty
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.CustomCollectors
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.NotFoundException
import org.springframework.context.ApplicationEventPublisher
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(rollbackFor = [Exception::class])
class SpecialtyServiceImpl(
    private val specialtyRepository: SpecialtyRepository,
    private val publisher: ApplicationEventPublisher,
    private val specialtyMapper: SpecialtyMapper,
) : SpecialtyService {

    override fun findAll(): List<SpecialtyDTO> {
        val specialties = specialtyRepository.findAll(Sort.by("id"))
        return specialties.map { specialty ->
            specialtyMapper.updateSpecialtyDTO(
                specialty,
                SpecialtyDTO(),
            )
        }
    }

    override fun `get`(id: Int): SpecialtyDTO = specialtyRepository.findById(id)
        .map { specialty -> specialtyMapper.updateSpecialtyDTO(specialty, SpecialtyDTO()) }
        .orElseThrow { NotFoundException() }

    override fun create(specialtyDTO: SpecialtyDTO): Int {
        val specialty = Specialty()
        specialtyMapper.updateSpecialty(specialtyDTO, specialty)
        return specialtyRepository.save(specialty).id!!
    }

    override fun update(id: Int, specialtyDTO: SpecialtyDTO) {
        val specialty = specialtyRepository.findById(id)
            .orElseThrow { NotFoundException() }
        specialtyMapper.updateSpecialty(specialtyDTO, specialty)
        specialtyRepository.save(specialty)
    }

    override fun delete(id: Int) {
        val specialty = specialtyRepository.findById(id)
            .orElseThrow { NotFoundException() }
        publisher.publishEvent(BeforeDeleteSpecialty(id))
        specialtyRepository.delete(specialty)
    }

    override fun getSpecialtyValues(): Map<Int, Int> = specialtyRepository.findAll(Sort.by("id"))
        .stream()
        .collect(CustomCollectors.toSortedMap(Specialty::id, Specialty::id))
}
